### PR TITLE
 conversions: [torch_dist --> torch] & [core --> HF]

### PR DESCRIPTION
**Pull request by AleHD**
Clean PR for #48, reverting my bad merge. Original description by @dhia680:

> This PR extends tools/checkpoint to support the conversion of a Megatron core checkpoint to a LlamaForCausalLM (HF).
Run tools/checkpoint/do-convert.sh with your updated paths/variables to perform both conversions.

### CODE REVIEW

**Comment by Alvorecer721** [comment]
Here are a few observations regarding your conversion script:

1. The [torch dist → torch conversion script](https://github.com/swiss-ai/Megatron-LM/blob/hf-conversion2/scripts/conversion/torchdist_2_torch.py) may have issues because it only adds a minimal number of arguments to `args_defaults`—both in this script and in your [run conversion script](https://github.com/swiss-ai/Megatron-LM/blob/hf-conversion2/scripts/conversion/run_torchdist-2-torch.sh)—compared to the official launch script. This difference causes the metadata stored in the torch checkpoint to deviate from that used during pretraining. The attached screenshot shows the messages printed after I added a few parameters to `args_defaults`.
image

2. Both the [tools/checkpoint/convert.py](https://github.com/swiss-ai/Megatron-LM/blob/eval/tools/checkpoint/convert.py) and your [save_llama_hf.py](https://github.com/swiss-ai/Megatron-LM/blob/eval/tools/checkpoint/saver_llama_hf.py) rely on the metadata within the torch checkpoint. Consequently, if the torch dist → torch conversion script uses default values, the HF checkpoint will inherit the incorrect configuration, potentially affecting both inference and evaluation.

3. I was unable to run your [torch dist → torch conversion script](https://github.com/swiss-ai/Megatron-LM/blob/hf-conversion2/scripts/conversion/torchdist_2_torch.py) because it repeatedly reported missing parameters. This makes me wonder how you managed to perform the conversion with the scripts you provided, did you forget to add some change? Could you please clarify it a bit, many thanks!

**Comment by dhia680** [comment]
The saver if purely for llama architecture so that might explain the missing params you got.
I or @AleHD will check this point of overriden args during torch dist conversion.
Thanks for the review !

**Comment by Alvorecer721** [comment]
> The saver if purely for llama architecture so that might explain the missing params you got. I or @AleHD will check this point of overriden args during torch dist conversion. Thanks for the review !

emm I am trying to convert a llama architecture

**Comment by dhia680** [comment]
What params are missing ?

**Comment by AleHD** [comment]
Thanks for the feedback! Regarding the torchdist->torch converter in particular, I'm unsure why you got so many arguments missing, I have never seen that before. In particular, at least the core architecture arguments (num layers, activation, hidden size, etc) are all loaded because of the `use_checkpoint_args`. Are you sure you specified the correct checkpoint location? Maybe it's missing.

Other than that, the more subtle but important arguments like rope theta, dropout, etc should be indeed double checked to ensure the conversion is correct, thanks for bringing that up!

**Comment by dhia680** [comment]
Hi @Alvorecer721 , could you please share what's needed to reproduce what you've initially encountered (e.g. exact commands, the ckpt you tested with...)

**Comment by Alvorecer721** [comment]
To reproduce the error:
```bash
CUDA_DEVICE_MAX_CONNECTIONS=1 torchrun src/convert/convert_torch_dist_to_torch.py \
    --bf16 \
    --load /iopsstor/scratch/cscs/xyixuan/Megatron-LM/logs/Meg-Runs/DenseGutenberg/llama3-1b-standard-80gbsz \
    --ckpt-convert-save /iopsstor/scratch/cscs/xyixuan/Megatron-LM/logs/Meg-Runs/DenseGutenberg/llama3-1b-standard-80gbsz
``` 
Error message:
image
This is just the first missing flag being printed. When I run the conversion, multiple missing flags appear. I've tried this at least 10 times. Only after adding all the required flags in my [script](https://github.com/Alvorecer721/PDM/blob/main/src/convert/convert_torch_dist_to_torch.py) does the conversion work correctly (with the right parameters and model size).

However, even then it's just initializing a new model as indicated by the converted folder name `iter000000` rather than loading the last checkpoint. This issue is resolved by append `checkpoints` [run conversion script](https://github.com/swiss-ai/Megatron-LM/blob/hf-conversion2/scripts/conversion/run_torchdist-2-torch.sh). 

Your script works fine with checkpoints moved by  @ischlag, which the `cp` happens inside `checkpoints` folder, but will never work for the actual cases.

Suggested fix:
1. Fix the checkpoint loading
```
CUDA_DEVICE_MAX_CONNECTIONS=1 torchrun $TORCH_DIST_SCRIPT \
    --bf16 \
    --load ${TORCHDIST_CKPT_PATH}/checkpoints \
    --ckpt-convert-save $TORCH_CKPT_SAVE_PATH
```
2. Add flags for important arguments like rope theta, dropout

**Comment by AleHD** [comment]
Hey, so I see the issue. As you said, the path you are specifying is missing the `checkpoints/`, which means that the `--load` is actually not finding the checkpoint. So the correct checkpoint to specify should already have the `checkpoints/` suffix. So without having to set any of the arguments you mentioned, this is already successful as it loads them from the checkpoint:

```
CUDA_DEVICE_MAX_CONNECTIONS=1 torchrun scripts/conversion/torchdist_2_torch.py --bf16 --load /iopsstor/scratch/cscs/xyixuan/Megatron-LM/logs/Meg-Runs/DenseGutenberg/llama3-1b-standard-80gbsz/checkpoints --ckpt-convert-save $SCRATCH/temps/convert/yixuan/torchnodist
```

This makes me think that you are actually converting a randomly initialized model, as it really shouldn't be able to find any checkpoint at that `--load` path. Try also setting `--exit-on-missing-checkpoint` to confirm this.

That said, we still have to take a look at the more tricky args like rope_theta and so, thanks!

**Comment by AleHD** [comment]
Given this discussion, I would suggest making the following changes:
- Remove `run_torchdist-2-torch.sh` entirely as it seems redundant to me given the .py is rather easy to use already + the error on the current example that is missing the `checkpoints/` in the script.
- Add `exit_on_missing_checkpoint = True` default in `convert_torch_dist_to_torch.py` so this confusion doesn't happen again.
- And ofc, double check what arguments are loaded from checkpoint to make sure all hyperparameters are convertly set (in https://github.com/swiss-ai/Megatron-LM/blob/cb916e2eaa92e1d8bd038a0c26d9c32d43880522/megatron/training/checkpointing.py#L999-L1060).

Also pinging @dhia680.

**Comment by AleHD** [comment]
This branch now contains all of the conversion utilities. Ready to review and merge :)

**Comment by AleHD** [comment]
Conversions are working perfectly:
- The `scripts/conversion/README.md` serves as a good entrypoint documenting the usage of the additions.
- Using `--test-logits` we can see that megatron Apertus1B agrees with its HF implementation:
  ```
  Converted model agrees on 99.73% of predictions
  Converted logits are close on 97.79% of values
  Max absolute difference: 0.19936037063598633
  Mean absolute difference: 0.001554669812321663
  Max relative difference: inf
  Mean relative difference (no inf): 0.007421422284096479
  Relative difference inf proportion: 2.379426933263318e-10
   ```
- Apertus70B can be converted succesfully, can't be tested with `--test-logits` as before but that already suggest a correct conversion, plus HF generations are coherent enough. This is the output of prompting Apertus 70B @ iteration 50k with "Lausanne in Switzerland is":
  ```
  <s>Lausanne in Switzerland is the capital of the canton of Vaud. It is located on the shores of Lake Geneva, in the French-speaking part of Switzerland. The city is known for its beautiful architecture, its cultural events and its proximity to the Alps. Lausanne is a popular tourist destination, with many visitors coming to enjoy its sights and attractions.\nLausanne is a city in Switzerland that is known for its beautiful architecture and its proximity to the Alps. The city is also home to a number of cultural events and festivals, including the annual Lausanne Jazz Festival and the Lausanne International Film Festival.</s>
  ```

Should be good to merge :)

**Review by dhia680 — CHANGES_REQUESTED** [review]
Nice work @AleHD !
But 1 question: Why do we still call the saver script `saver_llama_hf.py` ?
Can you rename it to `saver_swissai_hf.py` and adapt things accordingly ?
We ideally need two savers, one for Llama and one for us.

**Comment by AleHD** [comment]
Thanks for the comments @dhia680! I changed the saver name accordingly. However, I wonder if we really need to have two savers given that the HF implementation SwissAIForCausalLM can also be used to implement llama models with the appropiate flags (disabling xielu, etc) and the current saver already accounts for that.

**Review by dhia680 — APPROVED** [review]
LGTM
